### ISSUE TITLE
[Kinemation Error] ArgumentException: AsyncGPUReadback cannot use Temp memory as input since the result may only become available at an unspecified point in the future.

### ISSUE THREAD

**Issue by Steven9Smith** [Dreaming381/Latios-Framework#34]
Platform: Unity 2022.3.2f1
Packages:
{
  "dependencies": {
    "com.unity.2d.sprite": "1.0.0",
    "com.unity.2d.tilemap": "1.0.0",
    "com.unity.ads": "4.4.2",
    "com.unity.ai.navigation": "1.1.3",
    "com.unity.analytics": "3.8.1",
    "com.unity.audio.dspgraph": "0.1.0-preview.22",
    "com.unity.collab-proxy": "2.0.4",
    "com.unity.entities": "1.0.10",
    "com.unity.entities.graphics": "1.0.10",
    "com.unity.ide.rider": "3.0.21",
    "com.unity.ide.visualstudio": "2.0.18",
    "com.unity.ide.vscode": "1.2.5",
    "com.unity.physics": "1.0.10",
    "com.unity.purchasing": "4.8.0",
    "com.unity.render-pipelines.universal": "14.0.8",
    "com.unity.test-framework": "1.1.33",
    "com.unity.textmeshpro": "3.0.6",
    "com.unity.timeline": "1.7.5",
    "com.unity.ugui": "1.0.0",
    "com.unity.visualeffectgraph": "14.0.8",
    "com.unity.visualscripting": "1.8.0",
    "com.unity.xr.legacyinputhelpers": "2.1.10",
    "com.unity.modules.ai": "1.0.0",
    "com.unity.modules.androidjni": "1.0.0",
    "com.unity.modules.animation": "1.0.0",
    "com.unity.modules.assetbundle": "1.0.0",
    "com.unity.modules.audio": "1.0.0",
    "com.unity.modules.cloth": "1.0.0",
    "com.unity.modules.director": "1.0.0",
    "com.unity.modules.imageconversion": "1.0.0",
    "com.unity.modules.imgui": "1.0.0",
    "com.unity.modules.jsonserialize": "1.0.0",
    "com.unity.modules.particlesystem": "1.0.0",
    "com.unity.modules.physics": "1.0.0",
    "com.unity.modules.physics2d": "1.0.0",
    "com.unity.modules.screencapture": "1.0.0",
    "com.unity.modules.terrain": "1.0.0",
    "com.unity.modules.terrainphysics": "1.0.0",
    "com.unity.modules.tilemap": "1.0.0",
    "com.unity.modules.ui": "1.0.0",
    "com.unity.modules.uielements": "1.0.0",
    "com.unity.modules.umbra": "1.0.0",
    "com.unity.modules.unityanalytics": "1.0.0",
    "com.unity.modules.unitywebrequest": "1.0.0",
    "com.unity.modules.unitywebrequestassetbundle": "1.0.0",
    "com.unity.modules.unitywebrequestaudio": "1.0.0",
    "com.unity.modules.unitywebrequesttexture": "1.0.0",
    "com.unity.modules.unitywebrequestwww": "1.0.0",
    "com.unity.modules.vehicles": "1.0.0",
    "com.unity.modules.video": "1.0.0",
    "com.unity.modules.vr": "1.0.0",
    "com.unity.modules.wind": "1.0.0",
    "com.unity.modules.xr": "1.0.0"
  }
}

Error:

`ArgumentException: AsyncGPUReadback cannot use Temp memory as input since the result may only become available at an unspecified point in the future.
UnityEngine.Rendering.AsyncRequestNativeArrayData.CreateAndCheckAccess[T] (Unity.Collections.NativeArray`1[T] array) (at <c2d036c16ca64e0eb93703a3b13e733a>:0)
UnityEngine.Rendering.AsyncGPUReadback.RequestIntoNativeArray[T] (Unity.Collections.NativeArray`1[T]& output, UnityEngine.GraphicsBuffer src, System.Action`1[T] callback) (at <c2d036c16ca64e0eb93703a3b13e733a>:0)
Latios.Kinemation.Authoring.Systems.MeshDeformDataSmartBlobberSystem.OnUpdate () (at ./Packages/com.latios.latiosframework/Kinemation/Authoring/BakingSystems/MeshDeformDataSmartBlobberSystem.cs:156)
Unity.Entities.SystemBase.Update () (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/SystemBase.cs:418)
Unity.Entities.ComponentSystemGroup.UpdateAllSystems () (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/ComponentSystemGroup.cs:735)
UnityEngine.Debug:LogException(Exception)
Unity.Debug:LogException(Exception) (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/Stubs/Unity/Debug.cs:19)
Unity.Entities.ComponentSystemGroup:UpdateAllSystems() (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/ComponentSystemGroup.cs:740)
Unity.Entities.ComponentSystemGroup:OnUpdate() (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/ComponentSystemGroup.cs:693)
Unity.Entities.SystemBase:Update() (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/SystemBase.cs:418)
Unity.Entities.ComponentSystemGroup:UpdateAllSystems() (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/ComponentSystemGroup.cs:735)
Unity.Entities.ComponentSystemGroup:OnUpdate() (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/ComponentSystemGroup.cs:693)
Unity.Entities.SystemBase:Update() (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/SystemBase.cs:418)
Unity.Entities.ComponentSystemGroup:UpdateAllSystems() (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/ComponentSystemGroup.cs:735)
Unity.Entities.ComponentSystemGroup:OnUpdate() (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/ComponentSystemGroup.cs:693)
Unity.Entities.SystemBase:Update() (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/SystemBase.cs:418)
Unity.Entities.BakingUtility:PostprocessBake(World, BakingSettings, BakingSystem) (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities.Hybrid/Baking/BakingUtility.cs:148)
Unity.Entities.BakingUtility:BakeScene(World, Scene, BakingSettings, Boolean, IncrementalBakingChangeTracker) (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities.Hybrid/Baking/BakingUtility.cs:85)
Unity.Scenes.Editor.LiveConversionDiffGenerator:Bake(BakingFlags) (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Scenes.Editor/LiveConversion/LiveConversionDiffGenerator.cs:214)
Unity.Scenes.Editor.LiveConversionDiffGenerator:UpdateLiveConversion(Scene, Hash128, LiveConversionDiffGenerator&, LiveConversionMode, UInt64, GUID, IEntitiesPlayerSettings, LiveConversionChangeSet&) (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Scenes.Editor/LiveConversion/LiveConversionDiffGenerator.cs:447)
Unity.Scenes.Editor.LiveConversionConnection:AddLiveConversionChangeSet(LiveConversionDiffGenerator&, Hash128, List`1, UInt64, LiveConversionMode) (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Scenes.Editor/LiveConversion/LiveConversionConnection.cs:417)
Unity.Scenes.Editor.LiveConversionConnection:Update(List`1, NativeList`1, NativeList`1, LiveConversionMode) (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Scenes.Editor/LiveConversion/LiveConversionConnection.cs:382)
Unity.Scenes.Editor.EditorSubSceneLiveConversionSystem:OnUpdate() (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Scenes.Editor/EditorSubSceneLiveConversionSystem.cs:64)
Unity.Entities.SystemBase:Update() (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/SystemBase.cs:418)
Unity.Entities.ComponentSystemGroup:UpdateAllSystems() (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/ComponentSystemGroup.cs:735)
Unity.Entities.ComponentSystemGroup:OnUpdate() (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/ComponentSystemGroup.cs:693)
Unity.Scenes.Editor.LiveConversionEditorSystemGroup:OnUpdate() (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Scenes.Editor/LiveConversion/LiveConversionEditorSystemGroup.cs:15)
Unity.Entities.SystemBase:Update() (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/SystemBase.cs:418)
Latios.SuperSystem:UpdateManagedSystem(LatiosWorldUnmanagedImpl*, ComponentSystemBase) (at ./Packages/com.latios.latiosframework/Core/Framework/SuperSystem.cs:292)
Latios.SuperSystem:UpdateAllSystems(ComponentSystemGroup, SystemSortingTracker&) (at ./Packages/com.latios.latiosframework/Core/Framework/SuperSystem.cs:330)
Latios.Systems.LatiosInitializationSystemGroupManager:ShouldGroupUpdate(ComponentSystemGroup) (at ./Packages/com.latios.latiosframework/Core/Systems/_Essentials/LatiosWorldSystemGroups.cs:78)
Unity.Entities.ComponentSystemGroup:OnUpdate() (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/ComponentSystemGroup.cs:697)
Unity.Entities.InitializationSystemGroup:OnUpdate() (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/DefaultWorld.cs:169)
Unity.Entities.SystemBase:Update() (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/SystemBase.cs:418)
Unity.Entities.DummyDelegateWrapper:TriggerUpdate() (at ./Library/PackageCache/com.unity.entities@1.0.10/Unity.Entities/ScriptBehaviourUpdateOrder.cs:526)`

After going through the Kinemation setup i have come across this roadblock. I also loaded my project in a 2021 version of Unity and the error still persists. Any ideas?

This issue is triggered @ (at ./Packages/com.latios.latiosframework/Kinemation/Authoring/BakingSystems/MeshDeformDataSmartBlobberSystem.cs:156)

**Comment by Dreaming381** [comment]
It looks like Blend Shape baking broke.

If you can send me a repro project or perhaps even just the prefab that is causing problems, I will get this fixed.

Otherwise, the workaround is to remove blend shapes from your mesh.

**Comment by Steven9Smith** [comment]
Sorry for the slow response. I removed a bunch of licensed stuff from the project and used a fresh mixamo rig and no error occurs. Ill do some more testing and see if i can figure out the issue cause it seems to be the assets I'm using.

**Comment by Steven9Smith** [comment]
Ok so i managed to find some free model in the asset store that experience the same behavior. In the Project I tested 3 different models: "Simple" Model, "Complex" Model (Unity-Chan), and a Mixamo Model.  I noticed some burst leaks but that could just be unity's fault as not all the leak were from kinemation.  Here are some images and the project link is added below.

[Image1](https://github.com/Steven9Smith/Latios_Framework_GPU_Bug/blob/main/image1.png)

[GPU Error Image](https://github.com/Steven9Smith/Latios_Framework_GPU_Bug/blob/main/image2.png)

Test Project LInk: https://github.com/Steven9Smith/Latios_Framework_GPU_Bug

**Comment by Dreaming381** [comment]
Thanks for the amazing repro! This uncovered a bunch of edge cases that my existing tests didn't cover!

I have all the bugs I encountered fixed now and will release 0.7.5 in the next 24 hours. Once I've done that, there will be a few things you will need to do for your test scene to make things work.

![image](https://github.com/Dreaming381/Latios-Framework/assets/25555355/388f5c5c-d353-40dd-b33e-d967b59d48f1)
In this screenshot, anything that is white I switched to using a non-skinning material. Those are regular Mesh Renderers that are parented to bones rather than skinned, and consequently were reading from random skinning transforms or vertices. In 0.7.5 there's a warning if you accidentally add a deforming shader to a normal Mesh Renderer. As for the black materials, those I had to switch to use the LatiosDeform shader because they either required blend shapes or had more than 4 bone influences.

A 0.8 alpha will be released in a few days with the same fixes, but also compatible with Unity Transforms with the LATIOS_TRANSFORMS_UNITY scripting define.

I hope these new releases serve you much better! Let me know if you have any further questions or concerns!
